core/src/test/java/com/linecorp/armeria/client/proxy/ProxyClientIntegrationTest.java

```diff
@@ -218,6 +241,75 @@ void testHttpProxyBasicCase() throws Exception {
         assertThat(response.contentUtf8()).isEqualTo(SUCCESS_RESPONSE);
     }
 
+    @Test
+    void testHttpProxyUpgradeRequestFailure() throws Exception {
+        DYNAMIC_HANDLER.setChannelReadCustomizer((ctx, msg) -> {
+            if (!(msg instanceof FullHttpRequest)) {
+                ctx.close();
+            }
+            final HttpRequest request = (HttpRequest) msg;
+            final DefaultFullHttpResponse response;
+            if ("h2c".equals(request.headers().get(HttpHeaderNames.UPGRADE))) {
+                // reject http2 upgrade requests
+                final HttpHeaders headers = new DefaultHttpHeaders().add(CONNECTION, "close");
+                response = new DefaultFullHttpResponse(HTTP_1_1, NOT_IMPLEMENTED, EMPTY_BUFFER,
+                                                       headers, EmptyHttpHeaders.INSTANCE);
+            } else {
+                response = new DefaultFullHttpResponse(HTTP_1_1, HttpResponseStatus.OK,
+                                                       copiedBuffer("success", US_ASCII));
+            }
+            ctx.writeAndFlush(response);
+            ctx.close();
+        });
+
+        final ClientFactory clientFactory =
+                ClientFactory.builder().proxyConfig(ProxyConfig.connect(httpProxyServer.address()))
+                             .useHttp2Preface(false).build();
+        final WebClient webClient = WebClient.builder(SessionProtocol.HTTP, http1Server.endpoint())
+                                             .factory(clientFactory)
+                                             .decorator(LoggingClient.newDecorator())
+                                             .build();
+        final CompletableFuture<AggregatedHttpResponse> responseFuture =
+                webClient.get(PROXY_PATH).aggregate();
+        final AggregatedHttpResponse response = responseFuture.join();
+        assertThat(response.status()).isEqualByComparingTo(OK);
+        assertThat(response.contentUtf8()).isEqualTo(SUCCESS_RESPONSE);
```

Comment:
Could we also check if the client retried a `CONNECT` request, instead of `GET`? For example, you could send a response content that contains the request method instead of `success` above.

Reply:
I've changed the test so that it returns the requested method which is `GET`.
This check should ensure `CONNECT` requests were not made directly to the backend server.

I'm still brainstorming ways where we can ensure `CONNECT` was retried as you said.

Reply:
Decided to go with the "easy route" and just add a counter for the number of successful proxy requests.